timemachine/cpp/src/bound_potential.cu

```diff
@@ -58,15 +55,28 @@ void BoundPotential::execute_host(
     }
 };
 
+void BoundPotential::set_params(const std::vector<double> &params) {
+    if (params.size() > 0) {
+        if (params.size() * sizeof(*d_p.data) > d_p.size) {
+            throw std::runtime_error(
+                "parameter size is greater than device buffer size: " + std::to_string(params.size()) + " > " +
+                std::to_string(d_p.size / sizeof(*d_p.data)));
+        }
+        // NOTE: can't use d_p.copy_from here because new parameters size may be less than buffer size
```

Comment:
Shouldn't they always be the same size?

Reply:
Not necessarily, since we currently allow the updated parameters to be <= the buffer size. This is the behavior of `set_params_device` (which I assume is useful for Local MD), and I kept the same behavior for `set_params` for consistency.

Reply:
Hm - outside of local MD I'm not sure why this would be allowed, the vast majority of the time this would likely indicate a user error?

Reply:
It is useful for Local MD, though I was probably premature in recommending the exact same behavior as the device method. As when we do it in Local MD we also modify the underlying potential (https://github.com/proteneer/timemachine/blob/master/timemachine/cpp/src/local_md_potentials.cu#L246-L247) to match the shift in parameter size. Perhaps we should simply have `set_params` only allow the size to match exactly, which avoids an intermediate buffer which might allow us to speed up minimization.

Reply:
Changed to require `set_params` argument to have the same size as the buffer in https://github.com/proteneer/timemachine/pull/1119/commits/4f75acadfd7324d4416d6efac0ad1d59c3b8a8de.